### DISCUSSION TITLE
Firebase Dynamic Links params

### DISCUSSION THREAD

**Discussion by ilpwork** [invertase/react-native-firebase#4867] [Q&A] [answered]
How can i get a params from dynamic link such as utm_campaign utm_medium and utm_source in app? Is it even possible?

**Comment by mikehardy** [comment, marked as answer]
Could you provide links to the documentation related to your question that you have read?

**Reply by ilpwork** [reply]
Hi, i have read [this](https://firebase.google.com/docs/dynamic-links/analytics) at the end and [this](https://firebase.google.com/docs/dynamic-links/create-manually) analytics parameters

onLink and getInitialLink events only return minAppVersion and deep link

Also i tried to google my question but found nothing

I need this in order to track whether the user has registered after clicking on the ad link

I will be glad if you help me

**Reply by mikehardy** [reply]
As documented:

> This data can be viewed in the Dynamic Links section of the Firebase console and retrieved using a REST API.

So, you need to use the REST API. I have done something similar for dynamic links information that is unrelated, in my project. I made a cloud function that used the REST API and as part of the app process that required the extra information I called the cloud function, which retrieved the info I needed. Something like this

```typescript
mike@isabela:~/work/Kullki/ksocialscore/packages/firebase/src (rebranding) % more getProfileLinkStats.ts 
import * as functions from 'firebase-functions';
import * as googleapis from 'googleapis';
import { EventEmitter } from 'events';
import * as rp from 'request-promise-native';

import { processCorrelationIds, sendError, sendSuccess } from './lib';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const cors = require('cors')({ origin: true });

// Set up logging
// https://firebase.google.com/docs/functions/writing-and-viewing-logs

// https://us-central1-kscoreappdev.cloudfunctions.net/getPublicProfile
// http://localhost:5000/kscoreappdev/us-central1/getPublicProfile
export const getProfileLinkStats = functions
  .runWith({
    memory: functions.VALID_MEMORY_OPTIONS[1], // 128MB was attempted and results in out of memory
  })
  .https.onRequest(async (req, res) => {
    return cors(req, res, async () => {
      processCorrelationIds(req, res);

      if (!req.body.data || !req.body.data.link) {
        throw new functions.https.HttpsError('internal', 'Missing link');
      }

      const link = req.body.data.link;
      res.locals.link = link;

      const duration = req.body.data.duration ?? 7;

      console.log(
        `getProfileLinkStats for ${link} over ${duration} days, Correlation-ID: ${res.locals.correlationId}`
      );
      console.log('Full inbound message: ' + JSON.stringify(req.body.data));

      // const uid = 'dynamicLinkStatsChecker';
      let token = 'unknown';
      try {
        const auth = new googleapis.Auth.GoogleAuth({
          // Scopes can be specified either as an array or as a single, space-delimited string.
          scopes: ['https://www.googleapis.com/auth/firebase'],
        });
        const maybeToken = await auth.getAccessToken();
        if (!maybeToken) {
          throw new Error('Unable to get token');
        }
        token = maybeToken;

        console.log('got a token: ' + token);
      } catch (error) {
        console.log('Error creating custom token:', error);
      }

      const encodedShortLink = encodeURIComponent(link);
      const options = {
        method: 'GET',
        uri: `https://firebasedynamiclinks.googleapis.com/v1/${encodedShortLink}/linkStats?durationDays=${duration}`,
        gzip: true,
        json: true,
        headers: { Authorization: `Bearer ${token}` },
      };
      let body;
      try {
        body = await rp(options);
      } catch (e) {
        console.log('getProfileLinkStats error: ' + JSON.stringify(e));
        return sendError(res, e.error.error);
      }
      console.log('got link response: ' + JSON.stringify(body, null, 2));

      try {
        EventEmitter.defaultMaxListeners = 200;

        return sendSuccess(res, { result: body });
      } catch (e) {
        console.log('getProfileLinkStats failure', e);
        return sendError(res, 'Unable to getProfileLinkStats');
      }
    });
  });
```

What is confounding about using the REST API for dynamic links is that the token auth scopes are not automatically set up for you in the default firebase token you already have while operating in the firebase functions node environment, whereas they are for most things, so this code is pretty handy as it shows you how to get a special auth token that has the correct token scopes and works for this REST API - something you don't normally have to do.

You'll have to bend it to do the REST API call you want vs getting link stats.

This was all quite "fun" to figure out, I recommend reading this to fully understand it: https://github.com/firebase/firebase-admin-node/issues/111#issuecomment-699575434

**Reply by ilpwork** [reply]
Thank you for your reply

![image](https://user-images.githubusercontent.com/45626761/107015405-bf372f00-67ad-11eb-9d92-5bd1a9a68c17.png)

I need campaign parameters of events (utm_campaign, utm_medium and utm_source), I dont know what mean "custom" 😄 

However, if its true, how can i get dynamic link inside my app that user clicked for the REST API?

**Reply by mikehardy** [reply]
The documentation there is specific:
- any 'utm_' parameter is a "custom campaign attribution"
- no custom campaign attribution information will be available in firebase dynamic link analytics
- in fact - contrary to my code above, they are not even available via the REST API to dynamic links, only google analytics!

So I am going to say "not possible" except in an aggregate way, unless there is something in the google analytics REST API that could help you.

Perhaps what you really want to do is *also* encode the important parts of your utm_ attributes into the path of the link itself, that your app sees when the link is delivered? Like you would do for referral tracking (where the referrer information say a recommendation to install the app from User "ReferrerGuy1" to his friend would be `http://viralapp.example.com/invites/ReferrerGuy1`, then when the friend installs the app the link is delivered and you know who the invite came from, you might have your deep link itself be `http://viralapp.example.com/campaigns/id<utm_campaign_id_here>/media<utm_media_here>/source<utm_source_here>`

...when you see "campaigns" links, split the path components out, parse them and you'll know.

**Reply by ilpwork** [reply]
Thanks a lot for the answers!
